### PR TITLE
[SPARK-26356][SQL] remove SaveMode from data source v2

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

In data source v1, save mode specified in `DataFrameWriter` is passed to data source implementation directly, and each data source can define its own behavior about save mode. This is confusing and we want to get rid of save mode in data source v2.

For data source v2, we expect data source to implement the `TableCatalog` API, and end-users use SQL(or the new write API described in [this doc](https://docs.google.com/document/d/1gYm5Ji2Mge3QBdOliFV5gSPTKlX4q1DCBXIkiyMv62A/edit?ts=5ace0718#heading=h.e9v1af12g5zo)) to acess data sources. The SQL API has very clear semantic and we don't need save mode at all.

However, for simple data sources that do not have table management (like a JIRA data source, a noop sink, etc.), it's not ideal to ask them to implement the `TableCatalog` API, and throw exception here and there.

`TableProvider` API is created for simple data sources. It can only get tables, without any other table management methods. This means, it can only deal with existing tables.

`TableProvider` fits well with `DataStreamReader` and `DataStreamWriter`, as they can only read/write existing tables. However, `TableProvider` doesn't fit `DataFrameWriter` well, as the save mode requires more than just get table. More specifically, `ErrorIfExists` mode needs to check if table exists, and create table. `Ignore` mode needs to check if table exists. When end-users specify `ErrorIfExists` or `Ignore` mode and write data to `TableProvider` via `DataFrameWriter`, Spark fails the query and asks users to use `Append` or `Overwrite` mode.

The file source is in the middle of `TableProvider` and `TableCatalog`: it's simple but it can check table(path) exists and create table(path). That said, file source supports all the save modes.

Currently file source implements `TableProvider`, and it's not working because `TableProvider` doesn't support `ErrorIfExists` and `Ignore` modes. Ideally we should create a new API for path-based data sources, but to unblock the work of file source v2 migration, this PR proposes to special-case file source v2 in `DataFrameWriter`, to make it work.

This PR also removes `SaveMode` from data source v2, as now only the internal file source v2 needs it.

## How was this patch tested?

existing tests

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala [hunk 1]
```diff
@@ -367,31 +361,13 @@ class DataSourceV2Suite extends QueryTest with SharedSQLContext {
...
-          df.write.format(format).option("path", optionPath).save()
+          df.write.format(format).option("path", optionPath).mode("append").save()
           assert(!new File(sessionPath).exists)
           checkAnswer(spark.read.format(format).option("path", optionPath).load(), df)
         }
       }
     }
   }
-
-  test("SPARK-25700: do not read schema when writing in other modes except append and overwrite") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala]: This test is not needed anymore, as `ErrorIfExists` and `Ignore` modes are unsupported now.

**Comment by cloud-fan** [comment]
cc @rdblue @gengliangwang @gatorsmile

**Comment by cloud-fan** [comment]
retest this please

**Comment by gengliangwang** [comment]
+1 on the proposal.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/noop/NoopDataSource.scala [hunk 2]
```diff
@@ -46,9 +45,7 @@ private[noop] object NoopTable extends Table with SupportsWrite with SupportsStr
   override def capabilities(): util.Set[TableCapability] = Set(TableCapability.BATCH_WRITE).asJava
 }
 
-private[noop] object NoopWriteBuilder extends WriteBuilder
-  with SupportsSaveMode with SupportsTruncate {
-  override def mode(mode: SaveMode): WriteBuilder = this
+private[noop] object NoopWriteBuilder extends WriteBuilder with SupportsTruncate {
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/noop/NoopDataSource.scala]: Since now DSV2(except file sources) can only write to an existing table, here the write path of NoopDataSource will still fail (analyzer rule `ResolveOutputRelation`)

**Comment by gengliangwang** [comment]
retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 3]
```diff
@@ -376,14 +376,20 @@ trait V2WriteCommand extends Command {
...
   def outputResolved: Boolean = {
-    table.resolved && query.resolved && query.output.size == table.output.size &&
-        query.output.zip(table.output).forall {
-          case (inAttr, outAttr) =>
-            // names and types must match, nullability must be compatible
-            inAttr.name == outAttr.name &&
+    table.resolved && query.resolved && {
+      // When the table schema is empty, skip the schema check.
+      // TODO: introduce a table capability to disable schema check.
+      table.schema.isEmpty || {
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Why is this necessary for removing `SaveMode`? It seems unrelated.

I think that we should add the table capability instead of adding this check. What does this unblock?

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/TableProvider.java [hunk 4]
```diff
@@ -24,7 +24,7 @@
 
 /**
  * The base interface for v2 data sources which don't have a real catalog. Implementations must
- * have a public, 0-arg constructor.
+ * have a public, 0-arg constructor, and can only deal with existing tables.
```
Comment (rdblue) [inline sql/core/src/main/java/org/apache/spark/sql/sources/v2/TableProvider.java]: The phrase "deal with" is not very specific. I think it would be better to say what these tables can be used for: data operations like read, append, delete, and overwrite. Not operations that require metadata changes.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 5]
```diff
@@ -267,6 +266,23 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
 
       import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Implicits._
       provider.getTable(dsOptions) match {
+        // TODO: for backward compatibility reasons, the builtin file source needs to support all
+        // the save modes, which violates the semantic of `TableProvider`. Here we special-case
+        // file source and pass the save mode to file source directly. This hack can be removed
+        // after we figure out a general interface for path-based data sources.
+        case table: FileTable =>
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Why is this hack necessary? Why not put off v2 support for path-based tables?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala [hunk 6]
```diff
@@ -25,14 +25,13 @@ import org.apache.spark.{SparkEnv, SparkException, TaskContext}
...
 import org.apache.spark.rdd.RDD
-import org.apache.spark.sql.SaveMode
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions.Attribute
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
 import org.apache.spark.sql.execution.{SparkPlan, UnaryExecNode}
 import org.apache.spark.sql.sources.{AlwaysTrue, Filter}
 import org.apache.spark.sql.sources.v2.SupportsWrite
-import org.apache.spark.sql.sources.v2.writer.{BatchWrite, DataWriterFactory, SupportsDynamicOverwrite, SupportsOverwrite, SupportsSaveMode, SupportsTruncate, WriteBuilder, WriterCommitMessage}
+import org.apache.spark.sql.sources.v2.writer.{BatchWrite, DataWriterFactory, SupportsDynamicOverwrite, SupportsOverwrite, SupportsTruncate, WriteBuilder, WriterCommitMessage}
```
[the remark this answers is not in the document]
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: I don't agree with this.

Wildcard imports make it difficult to cherry-pick commits and increase conflicts. It is also difficult to see where symbols are coming from and pollutes the namespace with everything in a package instead of just the required names.

For example, I recently hit problems adding a `logical` package for expressions because of places that imported `expressions._` along with `plans._`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/noop/NoopSuite.scala [hunk 7]
```diff
@@ -54,7 +56,7 @@ class NoopSuite extends SharedSQLContext {
           accum.add(1)
           x
         }
-        .write.format("noop").save()
+        .write.format("noop").mode("append").save()
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/noop/NoopSuite.scala]: Can we change the default mode for v2 to `append` if no mode is set? That seems more reasonable than always requiring `mode("append")`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: The noop data source reports its schema as empty, but can accept input data set of any schema.

Previously, `SupportsSaveMode` was a special case, and we skip schema check for it. Now, only file source is a special case,  and schema check is applied to noop data source as well.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/noop/NoopSuite.scala (hunk 7 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/noop/NoopSuite.scala]: Personally I support it, but I'd like to do it in another PR, as we expect to have discussions about backward compatibility for changing the default mode.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Because we want to migrate file source to v2, to validate the API of `Table`, `ScanBuilder`, `WriteBuilder`, etc.

This hack is just to work around the issue that we do not have a proper entry API for path-based data source, which I believe we will have later on.

I think this is not a bad idea. We can unblock the file source migration, and we can keep the DS v2 clean (`FileTable` is an internal class). Besides that this hack will be removed once we have path-based API in DS v2.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 3 above)
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: It sounds like this commit depends on adding a capability to turn off schema validation. That commit should go in first to avoid adding code that we have to remember to revert later.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/noop/NoopSuite.scala (hunk 7 above)
Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/noop/NoopSuite.scala]: Okay, then I think this PR depends on the PR to make append the default mode for v2. Please get that in first.

Otherwise, this PR will make unnecessary changes to several places and we will have to roll those changes back to make append the default mode.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 5 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: File sources do not use v2 by default, so this is not a necessary change for this commit. I think it should be in a separate PR. We can discuss whether it is a good idea to add this hack in the next DSv2 sync.

Please remove it from this PR.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/noop/NoopSuite.scala (hunk 7 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/noop/NoopSuite.scala]: what if we end up with not changing the default mode (e.g. the community doesn't agree with it)? do you think that's a blocker to forbid `ErrorIfExists` and `Ignore` mode for `TableProvider`?

Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/noop/NoopSuite.scala]: No, I don't think so. We would just have to make changes like this one. My motivation is to avoid making these changes because this touches several files just to set the mode to append.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 5 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I'm adding this to my agenda for the sync-up.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/noop/NoopSuite.scala (hunk 7 above)
Comment (rxin) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/noop/NoopSuite.scala]: We can't change the default like this. It'd be a huge nightmare for users to upgrade, because it's a subtle semantic change. If we want to piss off all Spark users that write data pipelines, changing the default from one to another here is a great way to achieve that.

We can however introduce new APIs to make it more obvious what's happening.

Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/noop/NoopSuite.scala]: The proposal is only to change the default when using v2. If we don't change the default for v2, then the default is to always fail, which I think is worse.

**Comment by cloud-fan** [comment]
I'd like to move forward this PR with the schema check hack, because of https://github.com/apache/spark/pull/24469#issuecomment-491696938

In addition, this PR
1. makes append mode the default when writing to data source v2
2. keep the hack for file source. Previously the file source relied on `SupportsSaveMode` to support all the save modes. Several tests() rely on it and removing the hack fails these tests. If we want to remove the hack, we should do it in a new PR and update these tests.

What do you think? @rdblue

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 8]
```diff
@@ -56,13 +55,16 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
    * <li>`SaveMode.Overwrite`: overwrite the existing data.</li>
    * <li>`SaveMode.Append`: append the data.</li>
    * <li>`SaveMode.Ignore`: ignore the operation (i.e. no-op).</li>
-   * <li>`SaveMode.ErrorIfExists`: default option, throw an exception at runtime.</li>
+   * <li>`SaveMode.ErrorIfExists`: throw an exception at runtime.</li>
    * </ul>
+   * <p>
+   * When writing to data source v1, the default option is `ErrorIfExist`. When writing to data
+   * source v2, the default option is `Append`.
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I think @rxin objected to changing the default mode for DSv2. Has that changed?

I think that this is a good idea because the most sensible default for v2 is to append if no mode is set, instead of failing because the mode is not supported. This doesn't change v1 behavior, so I think it is okay.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 5 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I believe that we agreed in the sync-up to remove this hack. Is there a reason why it is still included?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala [hunk 9]
```diff
@@ -219,14 +219,14 @@ class DataSourceV2Suite extends QueryTest with SharedSQLContext {
         assert(spark.read.format(cls.getName).option("path", path).load().collect().isEmpty)
 
         spark.range(10).select('id as 'i, -'id as 'j).write.format(cls.getName)
-          .option("path", path).save()
+          .option("path", path).mode("append").save()
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala]: I don't think this change is needed because the default mode for v2 is append.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala]: Before my PR, we write the files twice: once without the save mode, once with append mode.

Now I switch order, to make sure that the second write doesn't specify save mode, and prove the default mode is append.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 8 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I think @rxin has a concern about breaking existing ETL pipelines. Now we only change the default mode for DS v2(exclude file sources), I think there is no compatibility issue.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I explained in the comment: https://github.com/apache/spark/pull/24233#issuecomment-492628836

Removing the hack breaks several tests and I'd like to do it in another PR. Since `SupportsSaveMode` is a hack, which means we already have a hack for file source v2 in `DataFrameWriter.save`. To keep the PR small, I think it's better to still keep the hack(it's not creating a new hack) and open a new PR to remove it and update the tests.

**Comment by cloud-fan** [comment]
Note that, file source v2 is still supported in `DataFrameWriter` after this PR. In a previous data source v2 community sync, we agreed to not support file source v2 in `DataFrameWriter`, until we implement a path-based data source v2 API.

I think this is out of the scope of this PR, which aims to remove the `SupportsSaveMode`. We should open another PR to remove the support of file source v2 in `DataFrameWriter`, and update the failed tests.

**Comment by rdblue** [comment]
@cloud-fan, I don't recall that conclusion from a sync. Can you quote from the notes that you're talking about?

I'm fine fixing this in a follow-up, as long as there's a blocker filed so that this doesn't go into the 3.0 release.

**Comment by cloud-fan** [comment]
Let me quote something from your meeting notes:
> PR #24233: Remove SaveMode
> Ryan: This PR has a hack that passes SaveMode directly to file sources. Because the behavior of v1 is unknown, this should not be included in a release.
> Wenchen: This is needed in order to use v2 file sources. Intent is migrate the file source to provide a reference implementation. Without the hack, v2 sources can’t be used (by default) because we know it [v2] is broken [i.e., can’t run CTAS].
> Ryan: v2 isn’t broken. The only problem is path-based tables because we don’t know how they should behave [Gengliang is looking into this]. Catalog tables aren’t broken so we should focus on getting that support done.
> Russel/Matt: is this needed for current tests to pass?
> Wenchen: no, tests run using v1.
> Russel: +1 for not including the hack.
> Ryan: Also, tests for the new plans will require an in-memory source that will be a cleaner reference implementation.

At that time, I didn't realize that we do have test cases relying on file source v2 writing. That's why I think it's better to do it in another PR and update the tests. I've created a ticket to track it: https://issues.apache.org/jira/browse/SPARK-27815

**Comment by rdblue** [comment]
@cloud-fan, I don't think that issue is accurate. The problem is that there is a hack to continue passing SaveMode into some sources. That's what needs to be removed. Please update the issue to be more clear and mark it as a blocker, then I think we can move forward with this PR.

**Comment by rdblue** [comment]
@cloud-fan, thanks for fixing the issue. I think this is ready to go. +1 from me.

**Comment by cloud-fan** [comment]
Hi @dongjoon-hyun , do you have time to take a look at this please? thanks!

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 10]
```diff
@@ -268,9 +269,26 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
 
       import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Implicits._
       provider.getTable(dsOptions) match {
+        // TODO: for backward compatibility reasons, the builtin file source needs to support all
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Can we have an IDed TODO? `TODO:` -> `TODO(SPARK-XXX):`?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 11]
```diff
@@ -268,9 +269,26 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
+        // the save modes, which violates the semantic of `TableProvider`. Here we special-case
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: nit. `Here we special-case`?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 12]
```diff
@@ -56,13 +55,16 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
    * <li>`SaveMode.Overwrite`: overwrite the existing data.</li>
    * <li>`SaveMode.Append`: append the data.</li>
    * <li>`SaveMode.Ignore`: ignore the operation (i.e. no-op).</li>
-   * <li>`SaveMode.ErrorIfExists`: default option, throw an exception at runtime.</li>
+   * <li>`SaveMode.ErrorIfExists`: throw an exception at runtime.</li>
    * </ul>
+   * <p>
+   * When writing to data source v1, the default option is `ErrorIfExist`. When writing to data
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: nit `ErrorIfExist` -> `ErrorIfExists`

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 13]
```diff
@@ -78,15 +80,14 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
    * @since 1.4.0
    */
   def mode(saveMode: String): DataFrameWriter[T] = {
-    this.mode = saveMode.toLowerCase(Locale.ROOT) match {
+    mode(saveMode.toLowerCase(Locale.ROOT) match {
       case "overwrite" => SaveMode.Overwrite
       case "append" => SaveMode.Append
       case "ignore" => SaveMode.Ignore
       case "error" | "errorifexists" | "default" => SaveMode.ErrorIfExists
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: This should be changed because "default" is not `ErrorIfExists` for DSv2.

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Could you add a test coverage for this? It seems that we don't have yet.

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @cloud-fan . Mostly, it looks good to me. I have only one concern about `.mode("default")` behavior, https://github.com/apache/spark/pull/24233#discussion_r287206462 and the other comments are minor.

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/TableProvider.java [hunk 14]
```diff
@@ -26,6 +26,9 @@
  * The base interface for v2 data sources which don't have a real catalog. Implementations must
  * have a public, 0-arg constructor.
  * <p>
+ * Note that, TableProvider can only apply data operations to existing tables, like read, append,
+ * delete, and overwrite. Not operations that require metadata changes, like create/drop tables.
```
Comment (gatorsmile) [inline sql/core/src/main/java/org/apache/spark/sql/sources/v2/TableProvider.java]: `Not operations` -> `It does not support the operations`

**Review by gatorsmile — COMMENTED** [review]
LGTM

**Comment by cloud-fan** [comment]
@dongjoon-hyun good catch about `.mode("default")`! I've fixed it and added tests.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 15]
```diff
@@ -80,14 +80,15 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
-      case "overwrite" => SaveMode.Overwrite
-      case "append" => SaveMode.Append
-      case "ignore" => SaveMode.Ignore
-      case "error" | "errorifexists" | "default" => SaveMode.ErrorIfExists
+    saveMode.toLowerCase(Locale.ROOT) match {
+      case "overwrite" => mode(SaveMode.Overwrite)
+      case "append" => mode(SaveMode.Append)
+      case "ignore" => mode(SaveMode.Ignore)
+      case "error" | "errorifexists" => mode(SaveMode.ErrorIfExists)
+      case "default" => this
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: +1

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 16]
```diff
@@ -239,15 +243,75 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext with Be
...
       .format("org.apache.spark.sql.test")
-      .mode(SaveMode.ErrorIfExists)
       .save()
     assert(LastOptions.saveMode === SaveMode.ErrorIfExists)
+
+    spark.range(10).write
+      .format("org.apache.spark.sql.test")
+      .mode("default")
+      .save()
+    assert(LastOptions.saveMode === SaveMode.ErrorIfExists)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: Thanks!

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM, too.
